File: Freedom/微信/基本类库/Common/WXSettingViewController.m
//  TLSettingViewController.m
//  Freedom
// Created by Super
#import "WXSettingViewController.h"
#import <UMMobClick/MobClick.h>
#import "WXModes.h"
@interface WechatSettingSwitchCell : UITableViewCell
@property (nonatomic, assign) id<WXSettingSwitchCellDelegate>delegate;
@property (nonatomic, strong) WXSettingItem *item;
@end
@interface WechatSettingButtonCell : UITableViewCell
@property (nonatomic, strong) WXSettingItem *item;
@end
@interface WXSettingCell : UITableViewCell
@property (nonatomic, strong) WXSettingItem *item;
@end
@interface WXSettingHeaderTitleView : UITableViewHeaderFooterView
@property (nonatomic, strong) UILabel *titleLabel;
@property (nonatomic, strong) NSString *text;
@end
@interface WechatSettingFooterTitleView : WXSettingHeaderTitleView
@end
@interface WechatSettingSwitchCell ()
@property (nonatomic, strong) UILabel *titleLabel;
@property (nonatomic, strong) UISwitch *cellSwitch;
@end
@implementation WechatSettingSwitchCell
- (id)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier{
    if (self = [super initWithStyle:style reuseIdentifier:reuseIdentifier]) {
        [self setSelectionStyle:UITableViewCellSelectionStyleNone];
        [self setAccessoryView:self.cellSwitch];
        [self.contentView addSubview:self.titleLabel];
        [self p_addMasonry];
    }
    return self;
}
- (void)setItem:(WXSettingItem *)item{
    _item = item;
    [self.titleLabel setText:item.title];
}
#pragma mark - Event Response -
- (void)switchChangeStatus:(UISwitch *)sender{
    if (_delegate && [_delegate respondsToSelector:@selector(settingSwitchCellForItem:didChangeStatus:)]) {
        [_delegate settingSwitchCellForItem:self.item didChangeStatus:sender.on];
    }
}
#pragma mark - Private Methods -
- (void)p_addMasonry{
    [self.titleLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.mas_equalTo(self.contentView);
        make.left.mas_equalTo(self.contentView).mas_offset(15);
        make.right.mas_lessThanOrEqualTo(self.contentView).mas_offset(-15);
    }];
}
#pragma mark - Getter -
- (UILabel *)titleLabel{
    if (_titleLabel == nil) {
        _titleLabel = [[UILabel alloc] init];
    }
    return _titleLabel;
}
- (UISwitch *)cellSwitch{
    if (_cellSwitch == nil) {
        _cellSwitch = [[UISwitch alloc] init];
        [_cellSwitch addTarget:self action:@selector(switchChangeStatus:) forControlEvents:UIControlEventValueChanged];
    }
    return _cellSwitch;
}
@end
@interface WXSettingCell ()
@property (nonatomic, strong) UILabel *titleLabel;
@property (nonatomic, strong) UILabel *rightLabel;
@property (nonatomic, strong) UIImageView *rightImageView;
@end
@implementation WXSettingCell
- (id) initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier{
    if (self = [super initWithStyle:style reuseIdentifier:reuseIdentifier]) {
        [self.contentView addSubview:self.titleLabel];
        [self.contentView addSubview:self.rightLabel];
        [self.contentView addSubview:self.rightImageView];
        [self p_addMasonry];
    }
    return self;
}
- (void) setItem:(WXSettingItem *)item{
    _item = item;
    [self.titleLabel setText:item.title];
    [self.rightLabel setText:item.subTitle];
    if (item.rightImagePath) {
        [self.rightImageView setImage: [UIImage imageNamed:item.rightImagePath]];
    }else if (item.rightImageURL){
        [self.rightImageView sd_setImageWithURL:[NSURL URLWithString:item.rightImageURL] placeholderImage:[UIImage imageNamed:PuserLogo]];
    }else{
        [self.rightImageView setImage:nil];
    }
    
    if (item.showDisclosureIndicator == NO) {
        [self setAccessoryType: UITableViewCellAccessoryNone];
        [self.rightLabel mas_updateConstraints:^(MASConstraintMaker *make) {
            make.right.mas_equalTo(self.contentView).mas_offset(-15.0f);
        }];
    }else{
        [self setAccessoryType:UITableViewCellAccessoryDisclosureIndicator];
        [self.rightLabel mas_updateConstraints:^(MASConstraintMaker *make) {
            make.right.mas_equalTo(self.contentView);
        }];
    }
    
    if (item.disableHighlight) {
        [self setSelectionStyle:UITableViewCellSelectionStyleNone];
    }else{
        [self setSelectionStyle:UITableViewCellSelectionStyleDefault];
    }
}
#pragma mark - Private Methods
- (void)p_addMasonry{
    [self.titleLabel setContentCompressionResistancePriority:500 forAxis:UILayoutConstraintAxisHorizontal];
    [self.titleLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.mas_equalTo(self.contentView);
        make.left.mas_equalTo(self.contentView).mas_offset(15);
    }];
    
    [self.rightLabel setContentCompressionResistancePriority:200 forAxis:UILayoutConstraintAxisHorizontal];
    [self.rightLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.right.mas_equalTo(self.contentView);
        make.centerY.mas_equalTo(self.titleLabel);
        make.left.mas_greaterThanOrEqualTo(self.titleLabel.mas_right).mas_offset(20);
    }];
    
    [self.rightImageView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.right.mas_equalTo(self.rightLabel.mas_left).mas_offset(-2);
        make.centerY.mas_equalTo(self.contentView);
    }];
}
#pragma mark - Getter
- (UILabel *)titleLabel{
    if (_titleLabel == nil) {
        _titleLabel = [[UILabel alloc] init];
    }
    return _titleLabel;
}
- (UILabel *)rightLabel{
    if (_rightLabel == nil) {
        _rightLabel = [[UILabel alloc] init];
        [_rightLabel setTextColor:[UIColor grayColor]];
        [_rightLabel setFont:[UIFont systemFontOfSize:15.0f]];
    }
    return _rightLabel;
}
- (UIImageView *)rightImageView{
    if (_rightImageView == nil) {
        _rightImageView = [[UIImageView alloc] init];
    }
    return _rightImageView;
}
@end
@implementation WechatSettingButtonCell
- (id) initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier{
    if (self = [super initWithStyle:style reuseIdentifier:reuseIdentifier]) {
        [self.textLabel setTextAlignment:NSTextAlignmentCenter];
    }
    return self;
}
- (void) setItem:(WXSettingItem *)item{
    _item = item;
    [self.textLabel setText:item.title];
}
@end
@implementation WechatSettingFooterTitleView
- (id) initWithReuseIdentifier:(NSString *)reuseIdentifier{
    if (self = [super initWithReuseIdentifier:reuseIdentifier]) {
        [self.titleLabel mas_remakeConstraints:^(MASConstraintMaker *make) {
            make.left.mas_equalTo(self.contentView).mas_offset(15);
            make.right.mas_equalTo(self.contentView).mas_offset(-15);
            make.top.mas_equalTo(self.contentView).mas_offset(5.0f);
        }];
    }
    return self;
}
@end
@implementation WXSettingHeaderTitleView
- (id) initWithReuseIdentifier:(NSString *)reuseIdentifier{
    if (self = [super initWithReuseIdentifier:reuseIdentifier]) {
        [self.contentView addSubview:self.titleLabel];
        [self.titleLabel mas_makeConstraints:^(MASConstraintMaker *make) {
            make.left.mas_equalTo(self.contentView).mas_offset(15);
            make.right.mas_equalTo(self.contentView).mas_offset(-15);
            make.bottom.mas_equalTo(self.contentView).mas_offset(-5.0f);
        }];
    }
    return self;
}
- (void) setText:(NSString *)text{
    _text = text;
    [self.titleLabel setText:text];
}
#pragma mark - Getter
- (UILabel *) titleLabel{
    if (_titleLabel == nil) {
        _titleLabel = [[UILabel alloc] init];
        [_titleLabel setTextColor:[UIColor grayColor]];
        [_titleLabel setFont:[UIFont systemFontOfSize:14.0f]];
        [_titleLabel setNumberOfLines:0];
    }
    return _titleLabel;
}
@end
@implementation WXSettingViewController
- (void)loadView{
    self.view = [[UIView alloc] initWithFrame:CGRectMake(0, 0, APPW, APPH)];
    self.tableView = [[UITableView alloc] initWithFrame:self.view.bounds style:UITableViewStyleGrouped];
    [self.tableView setTableHeaderView:[[UIView alloc] initWithFrame:CGRectMake(0, 0, APPW, 15.0f)]];
    [self.tableView setTableFooterView:[[UIView alloc] initWithFrame:CGRectMake(0, 0, APPW, 12.0f)]];
    [self.tableView setBackgroundColor:[UIColor lightGrayColor]];
    [self.tableView setLayoutMargins:UIEdgeInsetsMake(0, 15, 0, 0)];
    [self.tableView setSeparatorInset:UIEdgeInsetsMake(0, 15, 0, 0)];
    [self.tableView setSeparatorColor:[UIColor grayColor]];
}
- (void)viewDidLoad{
    [super viewDidLoad];
    
    [self.tableView registerClass:[WXSettingHeaderTitleView class] forHeaderFooterViewReuseIdentifier:@"TLSettingHeaderTitleView"];
    [self.tableView registerClass:[WechatSettingFooterTitleView class] forHeaderFooterViewReuseIdentifier:@"TLSettingFooterTitleView"];
    [self.tableView registerClass:[WXSettingCell class] forCellReuseIdentifier:@"TLSettingCell"];
    [self.tableView registerClass:[WechatSettingButtonCell class] forCellReuseIdentifier:@"TLSettingButtonCell"];
    [self.tableView registerClass:[WechatSettingSwitchCell class] forCellReuseIdentifier:@"TLSettingSwitchCell"];
}
- (void)viewWillAppear:(BOOL)animated{
    [super viewWillAppear:animated];
    [MobClick beginLogPageView:self.analyzeTitle];
}
- (void)viewWillDisappear:(BOOL)animated{
    [super viewWillDisappear:animated];
    [MobClick endLogPageView:self.analyzeTitle];
}
- (void)dealloc{
#ifdef DEBUG_MEMERY
    DLog(@"dealloc %@", self.navigationItem.title);
#endif
}
#pragma mark - Delegate -
//MARK: UITableViewDataSource
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView{
    return self.data.count;
}
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    return [(NSArray*)(self.data[section]) count];
}
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    WXSettingItem *item = [self.data[indexPath.section] objectAtIndex:indexPath.row];
    id cell = [tableView dequeueReusableCellWithIdentifier:item.cellClassName];
    [cell setItem:item];
    if (item.type == TLSettingItemTypeSwitch) {
        [cell setDelegate:self];
    }
    return cell;
}
//MARK: UITableViewDelegate
- (UIView *) tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section{
    WXSettingGroup *group = self.data[section];
    if (group.headerTitle == nil) {
        return nil;
    }
    WXSettingHeaderTitleView *view = [tableView dequeueReusableHeaderFooterViewWithIdentifier:@"TLSettingHeaderTitleView"];
    [view setText:group.headerTitle];
    return view;
}
- (UIView *) tableView:(UITableView *)tableView viewForFooterInSection:(NSInteger)section{
    WXSettingGroup *group = self.data[section];
    if (group.footerTitle == nil) {
        return nil;
    }
    WechatSettingFooterTitleView *view = [tableView dequeueReusableHeaderFooterViewWithIdentifier:@"TLSettingFooterTitleView"];
    [view setText:group.footerTitle];
    return view;
}
- (void) tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    [tableView deselectRowAtIndexPath:indexPath animated:NO];
}
- (CGFloat) tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath{
    return 44.0f;
}
- (CGFloat) tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section{
    WXSettingGroup *group = self.data[section];
    return 0.5 + (group.headerTitle == nil ? 0 : 5.0f + group.headerHeight);
}
- (CGFloat) tableView:(UITableView *)tableView heightForFooterInSection:(NSInteger)section{
    WXSettingGroup *group = self.data[section];
    return 20.0f + (group.footerTitle == nil ? 0 : 5.0f + group.footerHeight);
}
//MARK: TLSettingSwitchCellDelegate
- (void)settingSwitchCellForItem:(WXSettingItem *)settingItem didChangeStatus:(BOOL)on{
    [SVProgressHUD showErrorWithStatus:[NSString stringWithFormat:@"Switch事件未被子类处理Title: %@\nStatus: %@", settingItem.title, (on ? @"on" : @"off")]];
}
#pragma mark - Getter -
- (NSString *)analyzeTitle{
    if (_analyzeTitle == nil) {
        return self.navigationItem.title;
    }
    return _analyzeTitle;
}
@end

File: Freedom/微信/基本类库/Common/WXSettingViewController.swift
//
//  WXSettingViewController.swift
//  Freedom
import Foundation
import XExtension
import SnapKit
class WXSettingViewController: BaseTableViewController {
    var data: [WXSettingGroup] = []
    override func loadView() {
        view = UIView(frame: CGRect(x: 0, y: 0, width: APPW, height: APPH))
        tableView = UITableView(frame: view.bounds, style: .grouped)
        tableView.tableHeaderView = UIView(frame: CGRect(x: 0, y: 0, width: APPW, height: 15.0))
        tableView.tableFooterView = UIView(frame: CGRect(x: 0, y: 0, width: APPW, height: 12.0))
        tableView.backgroundColor = UIColor.lightGray
        tableView.layoutMargins = UIEdgeInsets(top: 0, left: 15, bottom: 0, right: 0)
        tableView.separatorInset = UIEdgeInsets(top: 0, left: 15, bottom: 0, right: 0)
        tableView.separatorColor = UIColor.gray
    }
    override func viewDidLoad() {
        super.viewDidLoad()
        tableView.register(WXSettingHeaderTitleView.self)
        tableView.register(WechatSettingFooterTitleView.self)
        tableView.register(WXSettingCell.self)
        tableView.register(WechatSettingButtonCell.self)
        tableView.register(WechatSettingSwitchCell.self)
    }

    override func numberOfSections(in tableView: UITableView) -> Int {
        return data.count
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return Int(data[section].items.count)
    }
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let group = data[indexPath.section]
        let item = group.items[indexPath.row]
        switch item.type {
        case .defalut1:
            let cell = tableView.dequeueCell(WXSettingCell.self)
            cell.viewModel.onNext(item)
            return cell
        case .titleButton:
            let cell = tableView.dequeueCell(WechatSettingButtonCell.self)
            cell.viewModel.onNext(item)
            return cell
        default:
            let cell = tableView.dequeueCell(WechatSettingSwitchCell.self)
            cell.viewModel.onNext(item)
            return cell
        }
    }
    override func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        let group: WXSettingGroup = data[section]
        if group.headerTitle.isEmpty {
            return nil
        }
        let view = tableView.dequeueHeadFootView(view: WXSettingHeaderTitleView.self)
        view?.text = group.headerTitle
        return view
    }

    override func tableView(_ tableView: UITableView, viewForFooterInSection section: Int) -> UIView? {
        let group: WXSettingGroup = data[section]
        if group.footerTitle.isEmpty {
            return nil
        }
        let view = tableView.dequeueHeadFootView(view: WechatSettingFooterTitleView.self)
        view?.text = group.footerTitle
        return view
    }
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: false)
    }
    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 44.0
    }
    override func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        let group: WXSettingGroup = data[section]
        return 0.5 + (group.headerTitle.isEmpty ? 0 : 5.0)
    }

    override func tableView(_ tableView: UITableView, heightForFooterInSection section: Int) -> CGFloat {
        let group: WXSettingGroup = data[section]
        return 20.0 + (group.footerTitle.isEmpty ? 0 : 5.0)
    }
}

class WechatSettingSwitchCell: BaseTableViewCell<WXSettingItem> {
    private var titleLabel = UILabel()
    private var cellSwitch = UISwitch()
    override func initUI() {
        selectionStyle = .none
        accessoryView = cellSwitch
        contentView.addSubview(titleLabel)
        viewModel.subscribe(onNext: {[weak self] (item) in
            guard let `self` = self else { return }
            self.titleLabel.text = item.title
        }).disposed(by: disposeBag)
        titleLabel.snp.makeConstraints { (make) in
            make.centerY.equalToSuperview()
            make.left.equalToSuperview().offset(15)
            make.right.lessThanOrEqualToSuperview().offset(-15)
        }
    }
}

class WechatSettingButtonCell: BaseTableViewCell<WXSettingItem> {
    override func initUI() {
        textLabel?.textAlignment = .center
        viewModel.subscribe(onNext: {[weak self] (item) in
            guard let `self` = self else { return }
            self.textLabel?.text = item.title
        }).disposed(by: disposeBag)
    }
}

class WXSettingCell: BaseTableViewCell<WXSettingItem> {
    private var titleLabel = UILabel()
    private var rightLabel = UILabel()
    private var rightImageView = UIImageView()

    override func initUI() {
        contentView.addSubview(titleLabel)
        contentView.addSubview(rightLabel)
        contentView.addSubview(rightImageView)
        rightLabel.textColor = UIColor.gray
        rightLabel.font = UIFont.systemFont(ofSize: 15.0)
        p_addMasonry()
        viewModel.subscribe(onNext: {[weak self] (item) in
            guard let `self` = self else { return }
            self.titleLabel.text = item.title
            self.rightLabel.text = item.subTitle
            if !item.rightImagePath.isEmpty {
                self.rightImageView.image = UIImage(named: item.rightImagePath)
            } else if !item.rightImageURL.isEmpty {
                self.rightImageView.sd_setImage(with: URL(string: item.rightImageURL), placeholderImage: Image.logo.image)
            } else {
                self.rightImageView.image = nil
            }
            if item.showDisclosureIndicator == false {
                self.accessoryType = .none
                self.rightLabel.snp.updateConstraints { (make) in
                    make.right.equalTo(self.contentView).offset(-15)
                }
            } else {
                self.accessoryType = .disclosureIndicator
                self.rightLabel.snp.makeConstraints { (make) in
                    make.right.equalTo(self.contentView)
                }
            }
            if item.disableHighlight {
                self.selectionStyle = .none
            } else {
                self.selectionStyle = .default
            }
        }).disposed(by: disposeBag)
    }
    func p_addMasonry() {

        titleLabel.snp.makeConstraints { (make) in
            make.centerY.equalToSuperview()
            make.left.equalToSuperview().offset(15)
        }
        rightLabel.snp.makeConstraints { (make) in
            make.right.equalToSuperview()
            make.centerY.equalToSuperview()
            make.left.greaterThanOrEqualTo(self.titleLabel.snp.right).offset(20)
        }
        rightImageView.snp.makeConstraints { (make) in
            make.right.equalTo(self.rightLabel.snp.left).offset(-2)
            make.centerY.equalToSuperview()
            make.size.equalTo(CGSize(width: 50, height: 50))
        }
    }
}

class WXSettingHeaderTitleView: UITableViewHeaderFooterView {
    lazy var titleLabel: UILabel = {
        let titleLabel = UILabel()
        titleLabel.textColor = UIColor.gray
        titleLabel.font = UIFont.systemFont(ofSize: 14.0)
        titleLabel.numberOfLines = 0
    return titleLabel
    }()
    var text = "" {
        didSet {
            titleLabel.text = text
        }
    }
    init(reuseIdentifier: String) {
        super.init(reuseIdentifier: reuseIdentifier)
        contentView.addSubview(titleLabel)
        titleLabel.snp.makeConstraints { (make) in
            make.left.equalTo(self.contentView).offset(15)
            make.right.equalTo(self.contentView).offset(-15)
            make.bottom.equalTo(self.contentView).offset(-5)
        }
    }
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

class WechatSettingFooterTitleView: WXSettingHeaderTitleView {
    override init(reuseIdentifier: String) {
        super.init(reuseIdentifier: reuseIdentifier)
        titleLabel.snp.makeConstraints { (make) in
            make.left.equalTo(self.contentView).offset(15)
            make.right.equalTo(self.contentView).offset(-15)
            make.top.equalTo(self.contentView).offset(5)
        }
    }
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}
